File: src/test/java/io/github/mschout/email/srs/provider/ShortCutSRSProviderTest.java
package io.github.mschout.email.srs.provider;

import static org.junit.jupiter.api.Assertions.*;

import com.google.common.collect.ImmutableList;
import io.github.mschout.email.srs.SRS;
import io.github.mschout.email.srs.SRSAddress;
import java.security.InvalidKeyException;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class ShortCutSRSProviderTest {

  @Test
  public void compileAndParse() throws InvalidKeyException {
    SRSProvider shortCutProvider = DefaultSRSProviderFactory
      .getInstance()
      .createProvider(SRS.Type.SHORTCUT, ImmutableList.of("dummy-secret"));

    String compiled = shortCutProvider.compile("example.com", "jdoe");

    assertTrue(compiled.endsWith("=example.com=jdoe"));
    assertTrue(compiled.startsWith("SRS0="));

    SRSAddress parsed = shortCutProvider.parse(compiled);

    assertNotNull(parsed);
    assertEquals(parsed.getHost(), "example.com");
    assertEquals(parsed.getUser(), "jdoe");
  }

  @Test
  public void conformanceTests() throws InvalidKeyException {
    // Ported from Perl Mail::SRS tests
    SRS srs = new SRS(SRS.Type.SHORTCUT, "foo");

    assertNotNull(srs);
    assertInstanceOf(SRS.class, (srs));
    assertEquals("foo", srs.getSecret());

    String source = "user@host.tld";

    List<String> aliases = IntStream
      .rangeClosed(0, 5)
      .boxed()
      .map(i -> String.format("alias%s@host%s.tld%s", i, i, i))
      .collect(Collectors.toList());

    String new0 = srs.forward(source, aliases.get(0));
    assertFalse(new0.isEmpty(), "Made a new address");
    assertTrue(new0.startsWith("SRS"), "It is an SRS address");
    String old0 = srs.reverse(new0);
    assertFalse(old0.isEmpty(), "Reversed the SRS address");
    assertEquals(old0, source, "The reversal was idempotent");

    String new1 = srs.forward(new0, aliases.get(1));
    assertFalse(new1.isEmpty(), "Made another new address with the SRS address");
    assertTrue(new1.startsWith("SRS"), "It is an SRS address");
    String old1 = srs.reverse(new1);
    assertFalse(old1.isEmpty(), "Reversed teh address again");
    assertEquals(old1, source, "Got back the original sender");

    List<String> tests = ImmutableList.of(
      "user@domain-with-dash.com",
      "user-with-dash@domain.com",
      "user+with+plus@domain.com",
      "user%with!everything&everything=@domain.somewhere"
    );

    String alias = "alias@host.com";

    for (String test : tests) {
      String srsaddr = srs.forward(test, alias);
      String oldaddr = srs.reverse(srsaddr);
      assertEquals(oldaddr, test, "Idempotent on " + test);
    }
  }
}

File: src/test/kotlin/io/github/mschout/email/srs/provider/ShortCutSRSProviderTest.kt
/*
 * Copyright 2026 Michael Schout
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.mschout.email.srs.provider

import io.github.mschout.email.srs.SRS
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldEndWith
import io.kotest.matchers.string.shouldNotBeEmpty
import io.kotest.matchers.string.shouldStartWith
import io.kotest.matchers.types.shouldBeInstanceOf

class ShortCutSRSProviderTest :
    FunSpec({
      test("compileAndParse") {
        val shortCutProvider =
            DefaultSRSProviderFactory.instance.createProvider(
                SRS.Type.SHORTCUT,
                listOf("dummy-secret"),
            )

        val compiled = shortCutProvider.compile("example.com", "jdoe")

        compiled shouldEndWith "=example.com=jdoe"
        compiled shouldStartWith "SRS0="

        val parsed = shortCutProvider.parse(compiled)

        parsed shouldNotBe null
        parsed.host shouldBe "example.com"
        parsed.user shouldBe "jdoe"
      }

      test("conformanceTests") {
        // Ported from Perl Mail::SRS tests
        val srs = SRS(SRS.Type.SHORTCUT, "foo")

        srs shouldNotBe null
        srs.shouldBeInstanceOf<SRS>()
        srs.secret shouldBe "foo"

        val source = "user@host.tld"

        val aliases = (0..5).map { i -> "alias$i@host$i.tld$i" }

        val new0 = srs.forward(source, aliases[0])
        new0.shouldNotBeEmpty()
        new0 shouldStartWith "SRS"
        val old0 = srs.reverse(new0)
        old0.shouldNotBeEmpty()
        old0 shouldBe source

        val new1 = srs.forward(new0, aliases[1])
        new1.shouldNotBeEmpty()
        new1 shouldStartWith "SRS"
        val old1 = srs.reverse(new1)
        old1.shouldNotBeEmpty()
        old1 shouldBe source

        val tests =
            listOf(
                "user@domain-with-dash.com",
                "user-with-dash@domain.com",
                "user+with+plus@domain.com",
                "user%with!everything&everything=@domain.somewhere",
            )

        val alias = "alias@host.com"

        for (test in tests) {
          val srsaddr = srs.forward(test, alias)
          val oldaddr = srs.reverse(srsaddr)
          oldaddr shouldBe test
        }
      }
    })
